fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/InMemoryQueueCursor.java

```diff
@@ -0,0 +1,207 @@
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.cursors;
+
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.RecordCursor;
+import com.apple.foundationdb.record.RecordCursorContinuation;
+import com.apple.foundationdb.record.RecordCursorProto;
+import com.apple.foundationdb.record.RecordCursorResult;
+import com.apple.foundationdb.record.RecordCursorVisitor;
+import com.apple.foundationdb.record.logging.LogMessageKeys;
+import com.apple.foundationdb.record.query.plan.plans.QueryResult;
+import com.apple.foundationdb.tuple.ByteArrayUtil2;
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.ZeroCopyByteString;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ForkJoinPool;
+
+/**
+ * A cursor that enables adding {@link QueryResult} elements to it. It stores these elements in a {@link Queue}
+ * allowing the consumers to retrieve them in the same order.
+ */
+public class InMemoryQueueCursor implements RecordCursor<QueryResult> {
+
+    @Nonnull
+    private final Executor executor;
+
+    @Nonnull
+    private final Queue<QueryResult> elements;
+
+    private boolean isClosed;
+
+    public InMemoryQueueCursor(@Nullable Continuation continuation) {
+        this(ForkJoinPool.commonPool(), continuation);
+    }
+
+    public InMemoryQueueCursor(@Nonnull final Executor executor) {
+        this(executor, null);
+    }
+
+    public InMemoryQueueCursor(@Nonnull final Executor executor, @Nullable Continuation continuation) {
+        this.executor = executor;
+        elements = new LinkedList<>();
+        if (continuation != null) {
+            elements.addAll(continuation.elements);
+        }
+        isClosed = false;
+    }
+
+    /**
+     * Add a new {@link QueryResult} element to the queue.
+     * @param element the new element to be added.
+     */
+    public void add(@Nonnull QueryResult element) {
+        elements.add(element);
+    }
+
+    /**
+     * Add a new {@link QueryResult} elements to the queue.
+     * @param elements the new elements to be added.
+     */
+    public void add(@Nonnull QueryResult... elements) {
+        Arrays.stream(elements).forEach(this::add);
+    }
+
+    @Nonnull
+    @Override
+    public CompletableFuture<RecordCursorResult<QueryResult>> onNext() {
+        if (!elements.isEmpty()) {
+            return CompletableFuture.completedFuture(RecordCursorResult.withNextValue(elements.poll(),
+                    new Continuation(ImmutableList.copyOf(elements))));
```

Comment:
Copying the elements here:
- it makes the `Continuation` state independent of the `Cursor`'s (with each invocation of `onNext` the cursor removes the head of the queue).
- it is more expensive, however it is a "true" snapshot, i.e. if the user adds more elements to the cursor after the continuation is created, these elements will not be part of the continuation, this seems different from the semantics of other continuations that store a mere offset for example, I am not sure if this is ok.

Having said that, should we change the design and make the cursor grow the added items without removing anything from it, this way, the continuation can be a mere index of the current head and we do not have to copy the elements.

Reply:
One of the things that needs to be considered is what to do about changes to the database between getting a continuation and using. it. For example, if the recursion is walking down a tree, what do we do if a row in the ancestor path gets deleted in the meantime? Do we continue to return some of its former descendants?